File: backend/Assessments/src/main/kotlin/assessments/domain/AssessmentRepository.kt
package org.example.assessments.domain

import org.example.templates.domain.TemplateDetailDto

interface AssessmentRepository {
    suspend fun create(req: CreateAssessmentRequest, adminId: Int, snapshot: TemplateDetailDto, templateName: String): Int
    suspend fun list(userId: Int, userRole: String, tab: String): List<AssessmentListItem>
    suspend fun tabCounts(userId: Int, userRole: String): AssessmentTabCounts
    suspend fun getById(id: Int): AssessmentDetail?

    suspend fun start(id: Int, userId: Int): TransitionResult
    suspend fun submit(id: Int, userId: Int): TransitionResult
    suspend fun beginReview(id: Int, userId: Int): TransitionResult
    suspend fun complete(id: Int, userId: Int, finalComment: String): TransitionResult

    suspend fun acquireLock(id: Int, userId: Int): LockResult
    suspend fun releaseLock(id: Int, userId: Int): Boolean

    suspend fun updateAnswers(id: Int, userId: Int, answers: List<AnswerDto>): UpdateResult
    suspend fun upsertComment(id: Int, userId: Int, itemId: Int, text: String): UpdateResult
}

File: web/src/assessments/domain/AssessmentRepository.ts
import type { AssessmentListItem, AssessmentTabCounts, AssessmentDetail, CreateAssessmentRequest, AnswerDto } from './types'

export interface AssessmentRepository {
  list(tab: string): Promise<AssessmentListItem[]>
  counts(): Promise<AssessmentTabCounts>
  getById(id: number): Promise<AssessmentDetail>
  create(req: CreateAssessmentRequest): Promise<number>
  start(id: number): Promise<void>
  submit(id: number): Promise<void>
  beginReview(id: number): Promise<void>
  complete(id: number, finalComment: string): Promise<void>
  acquireLock(id: number): Promise<void>
  releaseLock(id: number): Promise<void>
  updateAnswers(id: number, answers: AnswerDto[]): Promise<void>
  upsertComment(id: number, itemId: number, text: string): Promise<void>
}
